integration/oasis/stationgraph/station_graph_test.go

```diff
@@ -299,6 +423,100 @@ func testConnectivity(t *testing.T, graph *stationGraph, from string, tos []stri
+func TestGenerateChargeSolutions1(t *testing.T) {
+
+	fakeGraph1 := make([][]stationfinder.NeighborInfo, len(fakeNeighborsGraph))
+	for i := range fakeNeighborsGraph {
+		fakeGraph1[i] = make([]stationfinder.NeighborInfo, len(fakeNeighborsGraph[i]))
+		copy(fakeGraph1[i], fakeNeighborsGraph[i])
+	}
+
+	// generate channel contains neighbors information
+	// simulate real situation using different go-routine
+	c := make(chan stationfinder.WeightBetweenNeighbors)
+	go func() {
+		for _, n := range fakeGraph1 {
+			neighborsInfo := stationfinder.WeightBetweenNeighbors{
+				NeighborsInfo: n,
+				Err:           nil,
+			}
+			c <- neighborsInfo
+		}
+		close(c)
+	}()
+
+	currEnergyLevel := 20.0
+	maxEnergyLevel := 50.0
+	graph := NewStationGraph(c, currEnergyLevel, maxEnergyLevel,
+		chargingstrategy.NewFakeChargingStrategyCreator(maxEnergyLevel))
+	if graph == nil {
+		t.Error("create Station graph failed, expect none-empty graph but result is empty")
+	}
+
+	solutions := graph.GenerateChargeSolutions()
+	fmt.Printf("### %#v\n", solutions[0])
+	fmt.Printf("### %#v\n", solutions[0].ChargeStations[0])
+	fmt.Printf("### %#v\n", solutions[0].ChargeStations[1])
+	if len(solutions) != 1 {
+		t.Errorf("expect to have 1 solution but got %d.\n", len(solutions))
+	}
+	sol := solutions[0]
+	// 58.8 = 11.1 + 14.4 + 33.3
+	if !floatEquals(sol.Distance, 58.8) {
+		t.Errorf("Incorrect distance calculated for fakeGraph1 expect 58.89 but got %#v.\n", sol.Distance)
+	}
+
+	// 7918.8 = 11.1 + 2532(60% charge) + 14.4 + 5328(80% charge) + 33.3
+	if !floatEquals(sol.Duration, 7918.8) {
+		t.Errorf("Incorrect duration calculated for fakeGraph1 expect 10858.8 but got %#v.\n", sol.Duration)
+	}
+
+	// 6.7 = 40 - 33.3
+	if !floatEquals(sol.RemainingRage, 6.7) {
+		t.Errorf("Incorrect duration calculated for fakeGraph1 expect 10858.8 but got %#v.\n", sol.RemainingRage)
+	}
+
+	if len(sol.ChargeStations) != 2 {
+		t.Errorf("Expect to have 2 charge stations for fakeGraph1 but got %d.\n", len(sol.ChargeStations))
+	}
+
+	expectStation1 := &solution.ChargeStation{
+		Location: solution.Location{
+			Lat: 2.2,
+			Lon: 2.2,
+		},
+		StationID:     "station2",
+		ArrivalEnergy: 8.9,
+		WaitTime:      0,
+		ChargeTime:    2532,
+		ChargeRange:   30,
+	}
+	if !reflect.DeepEqual(sol.ChargeStations[0], expectStation1) {
+		t.Errorf("Expect first charge stations info for fakeGraph1 is %#v but got %#v\n", expectStation1, sol.ChargeStations[0])
+	}
+
+	expectStation2 := &solution.ChargeStation{
+		Location: solution.Location{
+			Lat: 5.5,
+			Lon: 5.5,
+		},
+		StationID:     "station5",
+		ArrivalEnergy: 15.6,
+		WaitTime:      0,
+		ChargeTime:    5328,
+		ChargeRange:   40,
+	}
+	if !reflect.DeepEqual(sol.ChargeStations[1], expectStation2) {
+		t.Errorf("Expect second charge stations info for fakeGraph1 is %#v but got %#v\n", expectStation2, sol.ChargeStations[1])
+	}
+
+}
+
 var epsilon float64 = 0.00000001
```

Comment:
Use `util.floatEquals` instead?

Reply:
> One more thing: it's better to always use `rebase` instead of `merge`, which can avoid strange commit like "@CodeBear801
> Merge branch 'feature/oasis-integration2-patch5' of [https://github.co…](https://github.co%E2%80%A6) …
> [c21093d](https://github.com/Telenav/osrm-backend/commit/c21093d96cd197cb93c571aa95d454349cebd8a6)
> …m/Telenav/osrm-backend into feature/oasis-integration2-patch5"

Ok, good suggestion.

Reply:
Recorded in https://github.com/Telenav/osrm-backend/issues/211
Will replace all of them together.